### DISCUSSION TITLE
Is it possible to set an `implicitNotFound` annotation inside a macro?

### DISCUSSION THREAD

**Discussion by soronpo** [scala/scala3#12434] [Metaprogramming]
In shapeless 2 there is this [macro trick](https://github.com/milessabin/shapeless/blob/d0a7562c49795f1ad6730e759c06cf4d7dccc763/core/src/main/scala/shapeless/lazy.scala#L224) for `Lazy` that enables to set the annotation of a symbol. I use this concept quite a lot in my own libraries to set proper error messages for implicit constraints. Is there a way to do this inside a Scala 3 macro?

**Comment by nicolasstucki** [comment]
What operation do you need to perform?

I see that we are missing the operation to add an annotation.

**Reply by soronpo** [reply]
My use-case is to modify the `implicitNotFound` annotation to customize the implicit error message.
In Scala 2 I used `c.enclosingImplicits` to get the symbol of last pending implicit, and I changed its `implicitNotFound` annotation to what I wanted. I created a custom `c.abort` extension method (see below) that always sets the annotation. 
So theoretically, in my use-case, I don't really need to do all of this if there was a way to get `report.error()` or `compile.error()` to do the same. Maybe add a function that does `injectImplicitNotFoundError`. 

```scala
  implicit class WhiteboxMacroExt(c : scala.reflect.macros.whitebox.Context) {
    import c.universe._
    private val defaultAnnotatedSym : Option[TypeSymbol] =
      if (c.enclosingImplicits.isEmpty) None else c.enclosingImplicits.last.pt match {
        case TypeRef(_,sym,_) => Some(sym.asType)
        case x => Some(x.typeSymbol.asType)
      }
    ////////////////////////////////////////////////////////////////////
    // Code thanks to Shapeless
    // https://github.com/milessabin/shapeless/blob/master/core/src/main/scala/shapeless/lazy.scala
    ////////////////////////////////////////////////////////////////////
    private def setAnnotation(msg: String, annotatedSym : TypeSymbol) : Unit = {
      import c.internal._
      import decorators._
      val tree0 =
        c.typecheck(
          q"""new _root_.scala.annotation.implicitNotFound("dummy")""",
          silent = false
        )

      class SubstMessage extends Transformer {
        val global = c.universe.asInstanceOf[scala.tools.nsc.Global]

        override def transform(tree: Tree): Tree = {
          super.transform {
            tree match {
              case Literal(Constant("dummy")) => Literal(Constant(msg))
              case t => t
            }
          }
        }
      }
      val tree = new SubstMessage().transform(tree0)
      annotatedSym.setAnnotations(Annotation(tree))
      ()
    }
    def setAnnotation(msg : String) : Unit = defaultAnnotatedSym.foreach(sym => setAnnotation(msg, sym))
    def abort(msg : String) : Nothing = {
      defaultAnnotatedSym.foreach(sym => setAnnotation(msg, sym))
      c.abort(c.enclosingPosition, msg)
    }
  }
```

**Reply by nicolasstucki** [reply]
What would the `injectImplicitNotFoundError` method do from the user perspective?

**Reply by soronpo** [reply]
Here is an example:
```scala
trait Foo[T]
given [T] (using Bar[T]) : Foo[T] = ???
trait Bar[T]
given [T] (using BiggerThan5[T]) : Bar[T] = ???
trait BiggerThan5[T]
inline given [T]  : BiggerThan5[T] = 
  inline compiletime.constValue[T] match 
    case i if i <= 5 => 
      compiletime.injectImplicitNotFounError("Argument is not bigger than 5")
      compiletime.error("Argument is not bigger than 5")
    case _ => new BiggerThan5[T]

val f1 = summon[Foo[3]] //expected error: Argument is not bigger than 5
val f2 = summon[Foo[6]] //OK
```
So `injectImplicitNotFounError` knows that the last implicit will be `Foo` and can modify its `implicitNotFound` accordingly. Or, we modify the error reported to consider such injected errors without modifying the `implicitNotFound` annotation. 
Maybe we can have an argument for `compiletime.error` instead, like `compiletime.error("Argument is not bigger than 5", injectImplicitError = true).

**Reply by nicolasstucki** [reply]
It seems that what we need is a mechanism to tell the type to not try more implicit searches.

**Reply by nicolasstucki** [reply]
Have you tried to return a `'{ compiletime.error(Argument is not bigger than 5) }`? This should tell the implicit search that it found something, but will later fail (as long as the implicit argument is not dropped).

**Reply by soronpo** [reply]
Hmmm... it surprised me that it works even without the quotes. I assumed Scala 2 behavior, but something has changed, I think. I'll do some more experimentation and get back to you.
